### PR TITLE
oauth2: Add default expiry for RFC compliance

### LINKED ISSUE #14542: [OAuth2] Incompatibility with access_token responses without "expire_in"
*Title*: *oauth2 filter is incompatible with authorization servers that omit `expire_in` on `access_token` responses*

*Description*:
When configuring the oauth2 HTTP filter against an RFC 6749-compliant OAuth 2.0 authorization server, the auth flow fails if the `access_token` response does not contain an `expire_in` field.

*Repro steps*:
Setup Envoy oauth2 HTTP filter against an RFC 6749-compliant authorization server that does not return the `expire_in` field on the `access_token` response, such as GitHub.

*Admin and Stats Output*: Not relevant

*Config*:
```yaml
http_filters:
- name: envoy.filters.http.oauth2
  typed_config:
    "@type": type.googleapis.com/envoy.extensions.filters.http.oauth2.v3alpha.OAuth2
    config:
      token_endpoint:
        cluster: github-oauth
        uri: https://github.com/login/oauth/access_token
        timeout: 3s
      authorization_endpoint: https://github.com/login/oauth/authorize
      redirect_uri: "https://example.org/callback" 
...
```

*Logs*:

Starting with the response headers of the `access_token` request as sent by the Envoy OAuth 2.0 client implementation. 

Relevant log line: `No access token or expiration after asyncGetAccessToken`

```
[2020-12-30 13:54:59.295][138392][debug][http] [source/common/http/async_client_impl.cc:99] async http request response headers (end_stream=false):                                                                 ':status', '200'                                                                                                                                                                                                    
'date', 'Wed, 30 Dec 2020 12:54:59 GMT'                                                                                                                                                                             'content-type', 'application/json; charset=utf-8'                                                                                                                                                                   
'transfer-encoding', 'chunked'                                                                                                                                                                                      
'server', 'GitHub.com'                                                                                                                                                                                              
'status', '200 OK'                                   
'vary', 'X-PJAX,Accept-Encoding, Accept, X-Requested-With,Accept-Encoding'                                
'etag', '<redacted>'       
'cache-control', 'max-age=0, private, must-revalidate'                                                    
'strict-transport-security', 'max-age=31536000; includeSubdomains; preload'                               
'x-frame-options', 'deny'                            
'x-content-type-options', 'nosniff'                  
'x-xss-protection', '1; mode=block'                  
'referrer-policy', 'origin-when-cross-origin, strict-origin-when-cross-origin'                            
'expect-ct', 'max-age=2592000, report-uri="https://api.github.com/_private/browser/errors"'                                                                                                                         
'content-security-policy', 'default-src 'none'; base-uri 'self'; block-all-mixed-content; connect-src 'self' uploads.github.com www.githubstatus.com collector.githubapp.com api.github.com github-cloud.s3.amazonaw
s.com github-production-repository-file-5c1aeb.s3.amazonaws.com github-production-upload-manifest-file-7fdce7.s3.amazonaws.com github-production-user-asset-6210df.s3.amazonaws.com cdn.optimizely.com logx.optimize
ly.com/v1/events wss://alive.github.com; font-src github.githubassets.com; form-action 'self' github.com gist.github.com; frame-ancestors 'none'; frame-src render.githubusercontent.com; img-src 'self' data: githu
b.githubassets.com identicons.github.com collector.githubapp.com github-cloud.s3.amazonaws.com *.githubusercontent.com; manifest-src 'self'; media-src 'none'; script-src github.githubassets.com; style-src 'unsafe
-inline' github.githubassets.com; worker-src github.com/socket-worker-5029ae85.js gist.github.com/socket-worker-5029ae85.js'
'x-github-request-id', 'BC8C:90EB:916225:CC179B:5FEC78A3'                                                 
'x-envoy-upstream-service-time', '134'               

[2020-12-30 13:54:59.295][138392][trace][http] [source/common/http/async_client_impl.cc:116] async http request response data (length=236 end_stream=false)
...
[2020-12-30 13:54:59.298][138392][debug][upstream] [source/extensions/filters/http/oauth2/oauth_client.cc:92] No access token or expiration after asyncGetAccessToken
[2020-12-30 13:54:59.298][138392][debug][http] [source/common/http/filter_manager.cc:827] [C1][S16538478899231496212] Sending local reply with details 
```

*Call Stack*: Not relevant

*Additional information*

[RFC 6749 Section 5.1](https://tools.ietf.org/html/rfc6749#section-5.1) states:

>   expires_in
>         RECOMMENDED.  The lifetime in seconds of the access token.  For
>         example, the value "3600" denotes that the access token will
>         expire in one hour from the time the response was generated.
>         If omitted, the authorization server SHOULD provide the
>         expiration time via other means or document the default value.

In short, authorization server implementations do not have to provide the `expires_in` parameter in their `access_token` request response.

For example, GitHub does not set this parameter: (https://docs.github.com/en/free-pro-team@latest/developers/apps/authorizing-oauth-apps)

```json
{
  "access_token":"e72e16c7e42f292c6912e7710c838347ae178b4a",
  "scope":"repo,gist",
  "token_type":"bearer"
}
```

The Envoy OAuth 2.0 client implementation interprets this field as mandatory though. It rejects any responses without the `expired_in` parameter

https://github.com/envoyproxy/envoy/blob/7e3b483bbb004bf2b8234e3ec2be6ceef254a2f9/source/extensions/filters/http/oauth2/oauth_client.cc#L91-L95

*Proposed Fix*

 #14625

The Envoy oauth2 filter token config should have a field `default_expires_in` or similar, that when specified, allows responses without the `expires_in` field to pass. In the absence of the `expires_in` field, it should use that default value from the config file as the token expiration parameter.

### PR DESCRIPTION
**Pull request by phlax**
Currently this filter does not work with Github - which would seem like one of the most likely use cases

This is a revival of https://github.com/envoyproxy/envoy/pull/14625

Should fix #14542

Unblocks #31534 

Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by phlax** [comment]
cc @terorie @williamsfu99 amsfu99

**Comment by phlax** [comment]
this appears to work as expected in local testing

File: api/envoy/extensions/filters/http/oauth2/v3/oauth.proto
Comment (adisuissa) [inline api/envoy/extensions/filters/http/oauth2/v3/oauth.proto:135]: Please add the default value if this field isn't set, and is not provided by the auth server.

**Review by adisuissa — COMMENTED** [review]
Left a small API comment.
cc @derekargueta @mattklein123 oauth2 code-owners.

File: api/envoy/extensions/filters/http/oauth2/v3/oauth.proto
Comment (phlax) [reply api/envoy/extensions/filters/http/oauth2/v3/oauth.proto:135]: iiuc you mean add here info about the default so it renders in docs - i think its 0s

> If this value is not set it will default to 0s and require for the expiry to be set by the server

?

Comment (phlax) [reply api/envoy/extensions/filters/http/oauth2/v3/oauth.proto:135]: updated on basis of my understanding

Comment (phlax) [reply api/envoy/extensions/filters/http/oauth2/v3/oauth.proto:135]: ive also added a related tip in the sandbox https://github.com/envoyproxy/envoy/pull/31534/files#diff-e4e2790b4749c030d10ae18ac6c9358096e08b1c6758ceca190eff113a4d13e0R199-R214

ill update the main oauth docs once that has landed

File: source/extensions/filters/http/oauth2/filter.h
Comment (adisuissa) [inline source/extensions/filters/http/oauth2/filter.h:162]: nit: no need to return a const

**Review by adisuissa — COMMENTED** [review]
Thanks!
/lgtm api

Assigning Matt as codeowner.
/assign @mattklein123

**Review by adisuissa — APPROVED** [review]
LGTM, thanks!

Oh, I think it needs a release note as well.

**Comment by phlax** [comment]
apologies for fp - im rebasing on this pr so it was to avoid the merge commit, pr commits were not changed

**Review by mattklein123 — APPROVED** [review]

**Pull request merged**
